src/binary.rs

```diff
@@ -0,0 +1,322 @@
+                version = version,
+                target = self.target()?
+            ),
+            Self::WasmOpt => format!(
+                "https://github.com/WebAssembly/binaryen/releases/download/{version}/binaryen-{version}-x86_64-{target}.tar.gz",
+                version = version,
+                target = self.target()?,
+            ),
+        })
+    }
+}
+
+/// Locate the given application and download it if missing.
+#[tracing::instrument(level = "trace")]
+pub async fn get(app: Application, version: Option<&str>) -> Result<PathBuf> {
+    let version = version.unwrap_or_else(|| app.default_version());
+
+    if let Ok(path) = find_system(app, version).await {
+        tracing::info!(app = app.name(), version = version, "using system installed binary");
+        return Ok(path);
+    }
+
+    let cache_dir = cache_dir()?;
+    let app_dir = cache_dir.join(format!("{}-{}", app.name(), version));
+    let bin_path = app_dir.join(app.path());
+
+    if !is_executable(&bin_path)? {
+        let path = download(app, version).await.context("failed downloading release archive")?;
+        let path2 = path.clone();
+
+        async_std::task::spawn_blocking(move || {
+            let mut file = File::open(path2).context("failed opening downloaded file")?;
+            install(app, &mut file, &app_dir)?;
+
+            Ok::<_, anyhow::Error>(())
+        })
+        .await?;
+
+        std::fs::remove_file(path).context("failed deleting temporary archive")?;
+    }
+
+    Ok(bin_path)
+}
+
+/// Try to find a globally system installed version of the application and ensure it is the needed
+/// release version.
+#[tracing::instrument(level = "trace")]
+async fn find_system(app: Application, version: &str) -> Result<PathBuf> {
+    let path = which::which(app.name())?;
+    let output = Command::new(&path).arg("--version").output().await?;
+
+    ensure!(output.status.success(), "running command failed");
+
+    let text = String::from_utf8_lossy(&output.stdout);
+    let text = text.trim();
+
+    let system_version = match app {
+        Application::WasmBindgen => text.splitn(2, ' ').nth(1).context("missing version")?.to_owned(),
+        Application::WasmOpt => text.splitn(2, ' ').nth(1).context("missing version")?.replace(' ', "_"),
+    };
+
+    if system_version == version {
+        Ok(path)
+    } else {
+        bail!("not found")
+    }
+}
+
+/// Check whether a given path exists, is a file and marked as executable (unix only).
+fn is_executable(path: &Path) -> Result<bool> {
+    if !path.exists() {
+        return Ok(false);
+    }
+
+    let metadata = path.metadata().with_context(|| anyhow!("failed getting metadata of {:?}", path))?;
+    if !metadata.is_file() {
+        return Ok(false);
+    }
+
+    #[cfg(unix)]
+    {
+        use std::os::unix::fs::PermissionsExt;
+        if metadata.permissions().mode() & 0o100 == 0 {
+            return Ok(false);
+        }
+    }
+
+    Ok(true)
+}
```

Comment:
I would say that this should probably be turned into an infallible fn. Any error conditions are semantically `false`. Thoughts?

Reply:
Initially I thought it would be helpful to get the details about why metadata couldn't be loaded so the user could fix wrong permissions to the folder for example.

For the cause of determining whether a binary is already available on the system, it doesn't really matter and we would fail in the try to download it anyways if folder permissions are missing.

So I'll make this infallible then 👍